### DISCUSSION TITLE
How to test "upload file" feature?

### DISCUSSION THREAD

**Discussion by VinodKumbar** [leancodepl/patrol#916] [Q&A] [answered]
### Discussed in https://github.com/leancodepl/patrol/discussions/629

<div type='discussions-op-text'>

<sup>Originally posted by **VinodKumbar** November 22, 2022</sup>
I got stuck here, how to automate this Upload Functionality ?

I want to automate the below actions 
1. Tap on Upload section ( this can be done using Integration driver or PatrolTest ) 
2. Select a file ( Native File Selection ) from Gallery / Drive (is there any option to native file selection in patrol Test ? )
 
</div>

**Comment by bartekpacia** [comment, marked as answer]
Hi @VinodKumbar 👋 Thanks for using Patrol!

It would be helpful if you shared the screenshots of the upload functionality that you want to test. Then I'll be able to help you :)

Regarding your question: please take a look at the [Inspecting native view hierarchy section](https://patrol.leancode.co/tips-and-tricks#inspecting-native-view-hierarchy) in our documentation.

**Comment by bartekpacia** [comment]
You might want to see #514 for an example test involving file picker.

**Comment by zyllus17** [comment]
Hi @bartekpacia I want to achieve the same test of uploading an image from gallery, can you help me out.

1.  Currently I am tapping the `Upload Image` button to be redirected to the Gallery. I am using `file_picker: ^5.2.5`. I am using the code below to do so.

```dart 
await $(t.upload_image).tap();
```
<img src="https://github.com/leancodepl/patrol/assets/121288318/29c74e2a-e9ad-484a-bcfc-723b574ec3e1" width="220" height="450">

 2. After that I am using this to open the pictures folder, as you suggested in #514 
```dart
  await $.native.tap(Selector(text: 'Pictures'));
```
<img src="https://github.com/leancodepl/patrol/assets/121288318/735c52e1-bf71-41c4-a17a-bf4dfbf29872" width="220" height="450">

3. Finally I want to select one image from inside this folder. In #514 you were able to see the name of the file but in my case there is no name displayed so Patrol isn't able to find it. 

**Can you tell me how I can select one image from the screen below?**

<img src="https://github.com/leancodepl/patrol/assets/121288318/c402cca5-e6b2-4d57-8c5e-6d622230cbcf" width="220" height="450">
